libgalois/src/GraphTopology.cpp

```diff
@@ -435,6 +435,182 @@ katana::EdgeTypeAwareTopology::MakeFrom(
       pg, edge_type_index, e_topo, std::move(per_type_adj_indices)});
 }
 
+std::unique_ptr<katana::ProjectedShuffleTopology>
+katana::ProjectedShuffleTopology::MakeProjectedGraph(
+    const katana::PropertyGraph* pg,
+    const std::vector<std::string>& node_properties,
+    const std::vector<std::string>& edge_properties) {
+  KATANA_LOG_DEBUG_ASSERT(pg);
+
+  const auto& topology = pg->topology();
+  if (topology.empty()) {
+    ProjectedShuffleTopology pst;
+    return std::make_unique<ProjectedShuffleTopology>(std::move(pst));
+  }
+
+  // calculate number of new nodes
+  uint32_t num_new_nodes = 0;
+  uint32_t num_new_edges = 0;
+
+  katana::DynamicBitset bitset_nodes;
+  bitset_nodes.resize(topology.num_nodes());
+
+  if (node_properties.empty()) {
+    num_new_nodes = topology.num_nodes();
+    // set all nodes
+    katana::do_all(katana::iterate(topology.all_nodes()), [&](auto src) {
+      bitset_nodes.set(src);
+    });
+  } else {
+    std::set<katana::EntityTypeID> node_entity_type_ids;
+
+    for (auto node_prop : node_properties) {
+      auto entity_type_id = pg->GetNodeEntityTypeID(node_prop);
+      node_entity_type_ids.insert(entity_type_id);
+    }
+
+    katana::GAccumulator<uint32_t> accum_num_new_nodes;
+    katana::GAccumulator<uint64_t> accum_num_new_edges;
+
+    katana::do_all(katana::iterate(topology.all_nodes()), [&](auto src) {
+      for (auto type : node_entity_type_ids) {
+        if (pg->DoesNodeHaveType(src, type)) {
```

Comment:
This also may be wrong, because if see the implementation of `DoesNodeHaveType`, it returns true if the type you provide is a subtype of the actual node type. You should be calling `GetTypeOfNode`. 

PS: as I look at our type interface, my head is starting to spin, so no surprise that you're confused about which methods to call.

Reply:
the `type` here is an atomic type. So I just want to check if it is a subtype of the actual node type. thanks

Reply:
This seems wrong. It goes back to my documentation question.

I think we want the projected nodes to have all the given types. If so, then get the entity type for the given set of strings/names. Then select nodes with that entity type (no for-loop in the operator).

Reply:
I can't find a function in the PropertyGraph.h file that takes a list of entity type names to return a entity type id. Thanks

Reply:
Use `GetNodeEntityTypeManager` and then get the type from a set of strings from it.